File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_SYSTEMINFO.cs
using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System;
using System.IO;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_SYSTEMINFO : Message
    {
        public SID_SYSTEMINFO()
        {
            Id = (byte)MessageIds.SID_SYSTEMINFO;
            Buffer = new byte[16];
        }

        public SID_SYSTEMINFO(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_SYSTEMINFO;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");

            if (context.Direction != MessageDirection.ClientToServer)
                throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} must be sent from client to server");

            if (Buffer.Length != 28)
                throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be 28 bytes");

            /**
             * (UINT32) Number of processors
             * (UINT32) Processor architecture
             * (UINT32) Processor level
             * (UINT32) Processor timing
             * (UINT32) Total physical memory
             * (UINT32) Total page file
             * (UINT32) Free disk space
             */

            using var m = new MemoryStream(Buffer);
            using var r = new BinaryReader(m);

            var cpuCount      = r.ReadUInt32();
            var cpuArch       = r.ReadUInt32();
            var cpuLevel      = r.ReadUInt32();
            var cpuTiming     = r.ReadUInt32();
            var totalRAM      = r.ReadUInt32();
            var totalSwap     = r.ReadUInt32();
            var freeDiskSpace = r.ReadUInt32();

            return true;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_SYSTEMINFO.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System
Imports System.IO

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_SYSTEMINFO
        Inherits Message

        Public Sub New()
            Id = CByte(MessageIds.SID_SYSTEMINFO)
            Buffer = New Byte(15) {}
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_SYSTEMINFO)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal context As MessageContext) As Boolean
            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
            If context.Direction <> MessageDirection.ClientToServer Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} must be sent from client to server")
            If Buffer.Length <> 28 Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be 28 bytes")

            Using m = New MemoryStream(Buffer)
                Using r = New BinaryReader(m)
                    Dim cpuCount = r.ReadUInt32()
                    Dim cpuArch = r.ReadUInt32()
                    Dim cpuLevel = r.ReadUInt32()
                    Dim cpuTiming = r.ReadUInt32()
                    Dim totalRAM = r.ReadUInt32()
                    Dim totalSwap = r.ReadUInt32()
                    Dim freeDiskSpace = r.ReadUInt32()
                End Using
            End Using

            Return True
        End Function
    End Class
End Namespace
